starky2/src/cross_table_lookup.rs

```diff
@@ -0,0 +1,299 @@
+    pub(crate) fn new(challenges: GrandProductChallengeSet<F>) -> Self {
+        Self {
+            challenges,
+            zs_columns: vec![],
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.zs_columns.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.zs_columns.is_empty()
+    }
+
+    pub fn z_polys(&self) -> Vec<PolynomialValues<F>> {
+        self.zs_columns.iter().map(|(p, _)| p.clone()).collect()
+    }
+}
+
+pub fn cross_table_lookup_data<F: RichField, C: GenericConfig<D, F = F>, const D: usize>(
+    config: &StarkConfig,
+    trace_poly_values: &[Vec<PolynomialValues<F>>],
+    cross_table_lookups: &[CrossTableLookup<F>],
+    challenger: &mut Challenger<F, C::Hasher>,
+) -> Vec<CtlData<F>> {
+    let challenges = get_grand_product_challenge_set(challenger, config.num_challenges);
+    cross_table_lookups.iter().fold(
+        vec![CtlData::new(challenges.clone()); trace_poly_values.len()],
+        |mut acc, cross_table_lookup| {
+            let CrossTableLookup {
+                looking_table,
+                looking_columns,
+                looked_table,
+                looked_columns,
+                default,
+            } = cross_table_lookup;
+
+            for &GrandProductChallenge { beta, gamma } in &challenges.challenges {
+                let z_looking = partial_products(
+                    &trace_poly_values[*looking_table as usize],
+                    looking_columns,
+                    beta,
+                    gamma,
+                );
+                let z_looked = partial_products(
+                    &trace_poly_values[*looked_table as usize],
+                    looked_columns,
+                    beta,
+                    gamma,
+                );
+
+                debug_assert_eq!(
+                    *z_looking.values.last().unwrap(),
+                    *z_looked.values.last().unwrap()
+                        * (gamma + reduce_with_powers(default.iter(), beta)).exp_u64(
+                            trace_poly_values[*looking_table as usize][0].len() as u64
+                                - trace_poly_values[*looked_table as usize][0].len() as u64
+                        )
+                );
+
+                acc[*looking_table as usize]
+                    .zs_columns
+                    .push((z_looking, looking_columns.clone()));
+                acc[*looked_table as usize]
+                    .zs_columns
+                    .push((z_looked, looked_columns.clone()));
+            }
+            acc
+        },
+    )
+}
+
+fn partial_products<F: Field>(
+    trace: &[PolynomialValues<F>],
+    columns: &[usize],
+    beta: F,
+    gamma: F,
+) -> PolynomialValues<F> {
+    let mut partial_prod = F::ONE;
+    let degree = trace[0].len();
+    let mut res = Vec::with_capacity(degree);
+    for i in 0..degree {
+        partial_prod *=
+            gamma + reduce_with_powers(columns.iter().map(|&j| &trace[j].values[i]), beta);
```

Comment:
I think it may be nice to have a combine method on `GrandProductChallenge`, since we do this in a few places, and it would save us from unpacking `GrandProductChallenge` into its parts

Reply:
Good idea! I've added it